extensions/sinks/redis/redis.go

```diff
@@ -0,0 +1,209 @@
+			r.sendSingle = i
+		}
+	}
+
+	r.dataType = "string"
+	if i, ok := props["dataType"]; ok {
+		if i, ok := i.(string); ok {
+			r.dataType = i
+		}
+	}
+
+	r.expiration = -1
+	if i, ok := props["expiration"]; ok {
+		if t, err := cast.ToInt(i, cast.STRICT); err == nil {
+			r.expiration = time.Duration(t)
+		}
+	}
+
+	return nil
+}
+
+func (r *RedisSink) Open(ctx api.StreamContext) (err error) {
+	logger := ctx.GetLogger()
+	logger.Debug("Opening redis sink")
+
+	r.cli = redis.NewClient(&redis.Options{
+		Addr:     r.addr,
+		Username: r.username,
+		Password: r.password,
+		DB:       r.db, // use default DB
+	})
+
+	return nil
+}
+
+func (r *RedisSink) Collect(ctx api.StreamContext, data interface{}) error {
+	logger := ctx.GetLogger()
+
+	if v, ok := data.([]byte); ok {
+		if r.field != "" {
+			if !r.sendSingle {
+				var out []map[string]interface{}
+				if err := json.Unmarshal(v, &out); err != nil {
+					logger.Debug("Failed to unmarshal data with error: ", err, " data:", string(v))
+					return err
+				}
+
+				for _, m := range out {
+					key := r.field
+					field, ok := m[key].(string)
+					if ok {
+						key = field
+					}
+
+					if r.dataType == "list" {
+						err := r.cli.LPush(ctx, key, v).Err()
+						if err != nil {
+							logger.Error(err)
+							return err
+						}
+						logger.Debugf("send redis list success, key:%s data: %s", key, string(v))
+					} else {
+						err := r.cli.Set(ctx, key, v, r.expiration*time.Second).Err()
+						if err != nil {
+							logger.Error(err)
+							return err
+						}
+						logger.Debugf("send redis string success, key:%s data: %s", key, string(v))
+					}
+				}
+			} else {
+				var out map[string]interface{}
+				if err := json.Unmarshal(v, &out); err != nil {
+					logger.Debug("Failed to unmarshal data with error: ", err, " data:", string(v))
+					return err
+				}
+				key := r.field
+				field, ok := out[key].(string)
+				if ok {
+					key = field
+				}
+
+				if r.dataType == "list" {
+					err := r.cli.LPush(ctx, key, v).Err()
+					if err != nil {
+						logger.Error(err)
+						return err
+					}
+					logger.Debugf("send redis list success, key:%s data: %s", key, string(v))
+				} else {
+					err := r.cli.Set(ctx, key, v, r.expiration*time.Second).Err()
+					if err != nil {
+						logger.Error(err)
+						return err
+					}
+					logger.Debugf("send redis string success, key:%s data: %s", key, string(v))
+				}
+			}
+
+		} else if r.key != "" {
+			if r.dataType == "list" {
+				err := r.cli.LPush(ctx, r.key, v).Err()
+				if err != nil {
+					logger.Error(err)
+					return err
+				}
+				logger.Debugf("send redis list success, key:%s data: %s", r.key, string(v))
+			} else {
+				err := r.cli.Set(ctx, r.key, v, r.expiration*time.Second).Err()
+				if err != nil {
+					logger.Error(err)
+					return err
+				}
+				logger.Debugf("send redis string success, key:%s data: %s", r.key, string(v))
+			}
+
+		} else {
+			logger.Debug("not config data key for redis")
```

Comment:
Is it better to validate this in configure?

Reply:
In redis JSON, the key is not optional, These are just records